File: src/main/java/net/sf/rails/game/model/ShareModel.java
package net.sf.rails.game.model;

import net.sf.rails.game.PublicCompany;

/**
 * ShareModel for displaying the share percentages
 */
public final class ShareModel extends RailsModel {
    
    private final PublicCompany company;

    private ShareModel(CertificatesModel parent, PublicCompany company) {
        super(parent, "shareModel_" + company.getId());
        this.company = company;
        // have share model observe floatation status of company
        company.getFloatedModel().addModel(this);
    }

    public static ShareModel create(CertificatesModel certModel, PublicCompany company) {
        ShareModel model = new ShareModel(certModel, company);
        certModel.addModel(model);
        return model;
    }
    
    @Override
    public CertificatesModel getParent() {
        return (CertificatesModel)super.getParent();
    }
    
    @Override
    public String toText() {
        return getParent().toText(company);
    }
    
}

File: dotnet/RailsLib.Net/Net/Game/Model/ShareModel.cs
using System;
using System.Collections.Generic;
using System.Text;

/**
 * ShareModel for displaying the share percentages
 */

namespace GameLib.Net.Game.Model
{
    public class ShareModel : RailsModel
    {
        private PublicCompany company;

        private ShareModel(CertificatesModel parent, PublicCompany company) : base(parent, "shareModel_" + company.Id)
        {
            this.company = company;
            // have share model observe floatation status of company
            company.GetFloatedModel().AddModel(this);
        }

        public static ShareModel Create(CertificatesModel certModel, PublicCompany company)
        {
            ShareModel model = new ShareModel(certModel, company);
            certModel.AddModel(model);
            return model;
        }

        new public CertificatesModel Parent
        {
            get
            {
                return (CertificatesModel)base.Parent;
            }
        }

        override public string ToText()
        {
            return Parent.ToText(company);
        }
    }
}
